Add language selector to support pre-login translations in HQ

nospame opened the pull request:
## Product Description
Adds this dropdown to the navbar pre-login to let a user select their language.
<img width="170" height="282" alt="image" src="https://github.com/user-attachments/assets/0d725e5c-ddee-4ffe-877a-b54ef0b845fa" />


## Technical Summary
https://dimagi.atlassian.net/browse/SAAS-18015
Adds a menu dropdown to the navbar when a user is not logged in, which lists the languages we currently have available in HQ as single-button forms that post a langcode to the `set_language` url.

Makes a built-in django i18n view `set_language` accessible in hqwebapp urls at ... `/set_language/`. If `update_django_locales` has never been run, as in my local environment prior to working on this feature, setting the language using one of our 3-letter langcodes (every language except English, Spanish, Swahili, and Afrikaans) will not have any effect on the page.

Also lets us use the `language_name_local` template filter with our custom langcodes by converting to 2-letter codes first.

Review by commit.

## Safety Assurance

### Safety story
Adds a new view based on `django.views.i18n.set_language` that sets the language and redirects based on form input. Uses a similar level of validation as that built-in view to ensure that: the referer we are redirecting to is safe to do so (if not, don't navigate anywhere), and that the langcode exists in `langcodes.langs_by_code` (otherwise, don't set it).

Tested locally and on staging. Also confirmed that if a user sets a language, then logs into an account that has a different language set, the account language takes precedence (by overwriting the `django_language` cookie).

### Automated test coverage
Added TestSetLanguage for the custom `set_language` view.

### QA Plan
Not planning it.


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Add builtin set_language to hqwebapp urls

nospame committed:
Add language selector to base_navigation.html

nospame committed:
Override language_name_local template filter

nospame committed:
Use language_name_local in navbar language selector

nospame committed:
isort imports

nospame committed:
Prettify base_navigation.html

nospame committed:
Bootstrap 5 Migration - Rebuilt diffs

biyeun commented on corehq/apps/hqwebapp/templates/hqwebapp/bootstrap5/base_navigation.html:
```diff
@@ -61,6 +61,30 @@
+              </button>
+              <ul class="dropdown-menu">
+                {% for lang in LANGUAGES %}
+                  <li>
```
i think bootstrap 5 handles dropdowns slightly differently... requiring a dropdown item class on the child elements
https://getbootstrap.com/docs/5.3/components/dropdowns/#single-button

nospame replied:
https://github.com/dimagi/commcare-hq/pull/36733/commits/7b63d5b8f9783d42865ff6a83c0b03e606229cf2

biyeun commented on corehq/apps/hqwebapp/templates/hqwebapp/bootstrap3/base_navigation.html:
```diff
@@ -44,18 +44,18 @@
-                        value="{{ lang.0 }}"
-                      >{{ lang.1 }}</button>
+                        value="{{ lang_code }}"
+                      >{{ lang_code|language_name_local|title }}</button>
```
suggestion: combine `language_name_local` and `title` since this is the only use-case?

biyeun reviewed (commented):
looks good so far, just bootstrap5 styles for dropdown and customizing set_language

nospame committed:
Define our own set_language

Validates against languages defined in langcodes, rather than django locales

nospame committed:
Title case output of custom language_name_local

nospame committed:
Update b5 base_navigation.html

nospame committed:
Bootstrap 5 Migration - Rebuilt diffs

nospame committed:
Validate next_url before using it to redirect

nospame commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -1543,3 +1545,27 @@
+    if lang_code and lang_code in langs_by_code.keys():
+        response.set_cookie(
+            settings.LANGUAGE_COOKIE_NAME,
+            lang_code,
```
Already validating with `if ... lang_code in langs_by_code.keys()`.

nospame commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -1543,3 +1545,27 @@ def check_sso_login_status(request):
+    next_url = request.META.get("HTTP_REFERER")
+    response = HttpResponseRedirect(next_url) if next_url else HttpResponse(status=204)
+    lang_code = request.POST.get("language")
+    if lang_code and lang_code in langs_by_code.keys():
```
One thing I'm not sure of is if this should validate against `langcodes` -- that is, if it's anything we consider a valid langcode even if we don't have translations for it -- or if it should validate against `settings.LANGUAGES`, which specifies languages we have translations for.

A user will only see the options in the dropdown that are defined in `settings.LANGUAGES` anyway, and if they were to mess with the inputs to enter set a different langcode, as long as it's valid according to `langcodes`, it will just set the cookie but they won't see the language change. So there's no functional difference, it's just which one feels like the right source of truth.

nospame replied:
I had a thought that it could be worth trying to fall back to the two letter langcode here if the locales haven't been updated, but after some additional fiddling with this, I don't think it is. Django doesn't seem to load translations for languages that aren't defined in `settings.LANGUAGES` even if they're present in config and file structure, which means that you can set that language cookie to whatever language you want, but if it's not in `settings.LANGUAGES`, django won't load in the translations for it.

This also helps with my source-of-truth dilemma -- going to use `LANGUAGES` because that's what django seems to be using.

nospame committed:
Get valid lang codes from settings.LANGUAGES

nospame committed:
Add TestSetLanguage

jingcheng16 commented on corehq/apps/hqwebapp/templatetags/hq_shared_tags.py:
```diff
@@ -600,6 +605,17 @@ def html_attr(value):
+    """
+    lang = langs_by_code.get(lang_code)
+    lang_code = lang['two'] if lang else ''
+    return i18n.language_name_local(lang_code).title()
```
It seems if a language only has three letter code, this will raise `KeyError: 'Unknown language code .'` Is this handled?

nospame replied:
This is a very good point, thank you. This shouldn't really happen in practice, but it is possible (and some languages in langcodes don't have two-letter versions, so would be calling `language_name_local` on an empty string that way). [0b87ac8](https://github.com/dimagi/commcare-hq/pull/36733/commits/0b87ac81e162bbb338de908965593e4e259e9516)

jingcheng16 reviewed (commented):
All new changes looks good to me except for `i18n.language_name_local('').title()` errors for me.

nospame committed:
Improve lang code handling in language_name_local

nospame commented on corehq/apps/hqwebapp/templatetags/hq_shared_tags.py:
```diff
@@ -612,8 +612,16 @@ def language_name_local(lang_code):
+            # fall back to English name from langcodes
+            lang_name = lang['names'][0]
+    else:
+        lang_name = ''
```
This shouldn't really happen in practice. But maybe it's better to fall back to just showing the langcode itself? I'm not sure.

jingcheng16 replied:
I think you're right about keeping close to the existing behavior! 😄

nospame committed:
Make custom language_name_local template filter behavior more like built-in

Meaning: it's okay to raise a KeyError if the lang code isn't valid, because this indicates an issue in programming and is not something a user could ever trigger.
